realm/realm-library/src/main/java/io/realm/RealmResults.java

```diff
@@ -56,22 +58,24 @@
  * @see Realm#executeTransaction(Realm.Transaction)
  */
 public class RealmResults<E extends RealmModel> extends OrderedRealmCollectionImpl<E> {
-    static <T extends RealmModel> RealmResults<T> createBacklinkResults(Realm realm, UncheckedRow row, Class<T> srcTableType, String srcFieldName) {
+    @SuppressLint("unused")
+    static <T extends RealmModel> RealmResults<T> createBacklinkResults(BaseRealm realm, Row row, Class<T> srcTableType, String srcFieldName) {
+        UncheckedRow uncheckedRow = (UncheckedRow) row;
         Table srcTable = realm.getSchema().getTable(srcTableType);
-        return new RealmResults<T>(
+        return new RealmResults<>(
                 realm,
-                Collection.createBacklinksCollection(realm.sharedRealm, row, srcTable, srcFieldName),
+                Collection.createBacklinksCollection(realm.sharedRealm, uncheckedRow, srcTable, srcFieldName),
                 srcTableType);
     }
 
-    static RealmResults<DynamicRealmObject> createBacklinkResults(DynamicRealm realm, CheckedRow row, Table srcTable, String srcFieldName) {
+    static RealmResults<DynamicRealmObject> createDynamicBacklinkResults(BaseRealm realm, Row row, Table srcTable, String srcFieldName) {
```

Comment:
The whole separation between DynamicRealm and Realm is pretty awful (and I wrote most of it) when it comes to our internal API's. I guess it boils down to two things: 1) DynamicRealmObject is a subclass of RealmObject which actually complicate things quite a lot 2) The difference between using Strings and Classes is permutating almost all code paths to the underlying code.

I wanted to revisit it for quite a while, but don't really have any good ideas on how to fix it.